### PR TITLE
Add new API to Set/Get/Delete the life cycle of the bucket

### PR DESCRIPTION
**Pull request by sinhaashish**
Add new API to Set/Get/Delete the life cycle of the bucket
New methods getBucketLifeCycle() and setBucketLifeCylce() introduced to get and set the lifeCycle of the bucket.

### CODE REVIEW

File: api/src/main/java/io/minio/MinioClient.java [hunk 1]
```diff
@@ -3939,6 +3940,88 @@ public void setBucketPolicy(String bucketName, String policy)
...
+          InsufficientDataException, IOException, InvalidKeyException, NoResponseException,
+          XmlPullParserException, ErrorResponseException, InternalException {
+    Map<String,String> queryParamMap = new HashMap<>();
+    queryParamMap.put("lifecycle", "");
+    byte[] buf = new byte[16384];
+    int bytesRead = 0;
+    HttpResponse response = null;
+    try {
+      response = executeGet(bucketName, "", null, queryParamMap);
+      bytesRead = response.body().byteStream().read(buf);
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MinioClient.java]: why do you read 16K only?

File: api/src/main/java/io/minio/MinioClient.java [hunk 2]
```diff
@@ -3939,6 +3940,88 @@ public void setBucketPolicy(String bucketName, String policy)
...
+      if (bytesRead < 0) {
+        // reached EOF
+        throw new IOException("reached EOF when reading bucket lifecycle");
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MinioClient.java]: Use `EOFException` or `MinioException` variant.

File: api/src/main/java/io/minio/MinioClient.java [hunk 3]
```diff
@@ -3939,6 +3940,129 @@ public void setBucketPolicy(String bucketName, String policy)
...
+   * @throws ErrorResponseException      upon unsuccessful execution
+   * @throws InternalException           upon internal library error
+   */
+  public void setBucketLifeCycle(String bucketName, String lifeCycle)
+          throws InvalidBucketNameException, NoSuchAlgorithmException,
+          InsufficientDataException, IOException, InvalidKeyException, NoResponseException,
+          XmlPullParserException, ErrorResponseException, InternalException {
+    // If lifecycle is empty or null then delete it
+    if (lifeCycle == null || "".equals(lifeCycle)) {
+      deleteBucketLifeCycle(bucketName);
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MinioClient.java]: You would need to error out or S3 service to error out, not to delete with `Set()`

File: api/src/main/java/io/minio/MinioClient.java [hunk 4]
```diff
@@ -3939,6 +3940,129 @@ public void setBucketPolicy(String bucketName, String policy)
...
+   * @throws InsufficientDataException   upon getting EOFException while reading given
+   *                                     InputStream even before reading given length
+   * @throws IOException                 upon connection error
+   * @throws InvalidKeyException         upon an invalid access key or secret key
+   * @throws NoResponseException         upon no response from server
+   * @throws XmlPullParserException      upon parsing response xml
+   * @throws ErrorResponseException      upon unsuccessful execution
+   * @throws InternalException           upon internal library error
+   */
+  private void deleteBucketLifeCycle(String bucketName)
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MinioClient.java]: make it `public`

File: api/src/main/java/io/minio/MinioClient.java (hunk 3 above)
Comment (sinhaashish) [reply api/src/main/java/io/minio/MinioClient.java]: done.

File: api/src/main/java/io/minio/MinioClient.java (hunk 4 above)
Comment (sinhaashish) [reply api/src/main/java/io/minio/MinioClient.java]: done.

**Comment by balamurugana** [comment]
Read [`createRequest()`](https://github.com/minio/minio-java/blob/master/api/src/main/java/io/minio/MinioClient.java#L925) why/how checksums are sent.

What happen if `Content-MD5` is not sent to AWS S3 in a valid case?

File: api/src/main/java/io/minio/MinioClient.java [hunk 5]
```diff
@@ -1035,6 +1035,7 @@ private HttpResponse execute(Method method, String region, String bucketName, St
 
     if (headerMap != null) {
       headerMap = normalizeHeaders(headerMap);
+
```
Comment (Praveenrajmani) [inline api/src/main/java/io/minio/MinioClient.java]: We can remove the extra empty lines in middle

File: api/src/main/java/io/minio/MinioClient.java [hunk 6]
```diff
@@ -3939,6 +3940,129 @@ public void setBucketPolicy(String bucketName, String policy)
...
+          throws InvalidBucketNameException, NoSuchAlgorithmException,
+          InsufficientDataException, IOException, InvalidKeyException, NoResponseException,
+          XmlPullParserException, ErrorResponseException, InternalException {
+    Map<String,String> queryParamMap = new HashMap<>();
+    queryParamMap.put("lifecycle", "");
+    HttpResponse response = executeDelete(bucketName,  "", queryParamMap);
+    response.body().close();
+  }
+
+
```
Comment (Praveenrajmani) [inline api/src/main/java/io/minio/MinioClient.java]: We can remove the extra empty lines in middle

File: docs/API.md [hunk 7]
```diff
@@ -554,6 +554,143 @@ try {
...
+                + "</Rule></LifecycleConfiguration>";
+
+
+  minioClient.setBucketLifecycle("lifecycleminiotest", lifeCycle);
+} catch (MinioException e) {
+  System.out.println("Error occurred: " + e);
+}
+```
+
+
```
Comment (Praveenrajmani) [inline docs/API.md]: We can remove the extra empty lines in middle

File: functional/FunctionalTest.java [hunk 8]
```diff
@@ -2636,6 +2636,72 @@ public static void setBucketPolicy_test1() throws Exception {
...
+      mintSuccessLog("getBucketLifeCycle(String bucketName)", null, startTime);
+    } catch (Exception e) {
+      mintFailedLog("getBucketLifeCycle(String bucketName) ",
+              null, startTime, null,
+              e.toString() + " >>> " + Arrays.toString(e.getStackTrace()));
+      throw e;
+    }
+  }
+
+
```
Comment (Praveenrajmani) [inline functional/FunctionalTest.java]: We can remove the extra empty lines in middle

File: functional/FunctionalTest.java [hunk 9]
```diff
@@ -2984,6 +3050,7 @@ public static void runTests() throws Exception {
     copyObject_test8();
     copyObject_test9();
 
+
```
Comment (Praveenrajmani) [inline functional/FunctionalTest.java]: We can remove the extra empty lines in middle

File: api/src/main/java/io/minio/MinioClient.java (hunk 5 above)
Comment (sinhaashish) [reply api/src/main/java/io/minio/MinioClient.java]: done

File: api/src/main/java/io/minio/MinioClient.java (hunk 6 above)
Comment (sinhaashish) [reply api/src/main/java/io/minio/MinioClient.java]: removed.

File: docs/API.md (hunk 7 above)
Comment (sinhaashish) [reply docs/API.md]: removed

File: functional/FunctionalTest.java (hunk 9 above)
Comment (sinhaashish) [reply functional/FunctionalTest.java]: done.

**Comment by sinhaashish** [comment]
> > Read [`createRequest()`](https://github.com/minio/minio-java/blob/master/api/src/main/java/io/minio/MinioClient.java#L925) why/how checksums are sent.
> > What happen if `Content-MD5` is not sent to AWS S3 in a valid case?
> 
> It responds with
> 
> ```
> ErrorResponse(code=InvalidRequest, message=Missing required header for this request: Content-MD5, bucketName=null, objectName=null, resource=null, requestId=39EAD8DE6656F92E, hostId=170cCvjHSCO2GL3q9unepDINvm7prhEIfX2W3kWeFJcIl96x2dE12LkIOhWObi5AnuHqfaDm7so=)
> request={method=PUT, url=http://praveenrajmani.s3.amazonaws.com/?lifecycle=, headers=x-amz-meta-Content-Length: 166
> Host: praveenrajmani.s3.amazonaws.com
> User-Agent: Minio (amd64; amd64) minio-java/dev
> x-amz-content-sha256: db794c8a32aed1c0889fbf76022496b00687645a7e2a6dfb041c3924ce5bebcb
> x-amz-date: 20190104T184115Z
> Authorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20190104/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date;x-amz-meta-content-length, Signature=*REDACTED*
> }
> response={code=400, headers=x-amz-request-id: 39EAD8DE6656F92E
> x-amz-id-2: 170cCvjHSCO2GL3q9unepDINvm7prhEIfX2W3kWeFJcIl96x2dE12LkIOhWObi5AnuHqfaDm7so=
> Content-Type: application/xml
> Transfer-Encoding: chunked
> Date: Fri, 04 Jan 2019 18:41:15 GMT
> Connection: close
> Server: AmazonS3
> }
> ```
> when `Content-MD5` is not sent in the request

@Praveenrajmani  .In the example you share you are using an  insecure connection i.e. `http` for amazon s3 and thus `Content-MD5` is not sent. I changed it to `https` and it worked.

**Comment by sinhaashish** [comment]
> Read [`createRequest()`](https://github.com/minio/minio-java/blob/master/api/src/main/java/io/minio/MinioClient.java#L925) why/how checksums are sent.
> 
> What happen if `Content-MD5` is not sent to AWS S3 in a valid case?



> Read [`createRequest()`](https://github.com/minio/minio-java/blob/master/api/src/main/java/io/minio/MinioClient.java#L925) why/how checksums are sent.
> 
> What happen if `Content-MD5` is not sent to AWS S3 in a valid case?

@balamurugana [PR](https://github.com/minio/minio-java/pull/569)  . With this PR  `Content-MD5` is not calculated  and thus it is not a part of request in `setBucketLifeCycle` for `http`.  While it works for `https`.   Should `Content-MD5` be added for all?

**Pull request merged**